Deprecate deterministic bucket_id and file_id

braydonf opened the pull request:
**Motivation**:
- Will give historically unique entry id if the same bucket entry name is used within the same bucket over time, and helps simplify billing logic.
- The same encryption key and iv will not be reused if the same bucket entry name is used w/ a different file, better security practices.
- Restores the ability to rename buckets and bucket entries *(file_id, bucket_id, and encryption_key are no longer derived from the name)*
- Generally helps to uniquely identify bucket entries

**Upgrading**:
- **The bucket entry id will *initially* remain as-is, and will be deprecated in the next major release. This is to introduce bucket entry `index` before deprecation.**
- Backwards compatible for downloads as `file_id` or `index` can be used to determine derivation of encryption key for the file. Those with the `index` field will use the new method.
- Older clients will need to upgrade to upload new files, an error is given to indicate that this is necessary to avoid any confusion.

**Other Notes**:
- Bucket entry creation can be made idempotent, such that it will be possible to detect if there is a duplicate, based on unique key `index`, and to not repeat the same action (e.g. storage events).
~~- Adds index to buckets for `user:name` to keep uniqueness of bucket names and to be able to query by bucket name efficiently~~
~~- Adds index to bucket entries for `bucket:name` to keep uniqueness of bucket entry names within a bucket~~

**Related**:
- https://github.com/Storj/bridge/issues/405
- https://github.com/Storj/bridge/issues/406

**Todo**:
- [x] Update clients to use bucket entry `index` when deriving encryption key https://github.com/Storj/libstorj/pull/286

Commit:
Bucket id based on frame id

- The frame id is required to be able to be created before any shard can be uploaded an can be used to deterministically derive an encryption key before the bucket entry is created
- Restores the ability to rename bucket entries
- Will give historically unique entry if the same bucket name is used within the same bucket, and simplifies billing logic.
- The same encryption key will not be reused if the same bucket name is used, better security practices.
- Is backwards compatible as existing bucket entry ids do not need to change and downloads will still be able to decrypt correctly, uploads will need to upgrade to the latest version that will create the encryption key based on frame_id instead of file_id.

Commit:
Introduce `index` on bucket entry

- Better user experience to indicate that clients need to upgrade when uploading
- Can remain backwards compatible on downloads w/ check for the existence of `index`

Commit:
Remove deterministic bucket id

Commit:
Ensure bucket names are unique

frdwrd commented:
How will keys be generated after this change?

braydonf commented:
For the most part in the same way, however it will use `index` instead of `file_id`.

frdwrd commented:
Okay cool. Please make sure the file encryption doc gets updated.
https://github.com/Storj/core/blob/master/doc/file-encryption.md

Commit:
Unique names per bucket

Commit:
Introduce alternative before deprecation

Commit:
Deprecate in the next release, introduce bucket entry `index` first

Commit:
Update test so that bucket entries can not be replaced

Commit:
Add deprecation message to tests

Commit:
Remove unique indexes for `user:bucket-name` and `bucket-id:file-name`

- There are currently duplicates in the database and would require a migration to update these to add the index
- It's still possible to have duplicate names as the encrypted versions could be different
- It's an application side issue that would need to be handled
